src/diffusers/pipelines/transformers_loading_utils.py

```diff
@@ -0,0 +1,94 @@
+if TYPE_CHECKING:
+    from transformers import PreTrainedModel, PreTrainedTokenizer
+
+if is_transformers_available():
+    from transformers import PretrainedConfig, PreTrainedModel, PreTrainedTokenizer
+
+if is_safetensors_available():
+    import safetensors.torch
+
+
+def load_tokenizer_from_dduf(
+    cls: "PreTrainedTokenizer", name: str, dduf_entries: Dict[str, DDUFEntry]
+) -> "PreTrainedTokenizer":
+    """
+    Load a tokenizer from a DDUF archive.
+
+    In practice, `transformers` do not provide a way to load a tokenizer from a DDUF archive. This function is a workaround
+    by extracting the tokenizer files from the DDUF archive and loading the tokenizer from the extracted files. There is an
+    extra cost of extracting the files, but of limited impact as the tokenizer files are usually small-ish.
+    """
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        for entry_name, entry in dduf_entries.items():
+            if entry_name.startswith(name + "/"):
+                tmp_entry_path = os.path.join(tmp_dir, *entry_name.split("/"))
+                with open(tmp_entry_path, "wb") as f:
+                    with entry.as_mmap() as mm:
+                        f.write(mm)
+        return cls.from_pretrained(tmp_dir, **kwargs)
+
+
+def load_transformers_model_from_dduf(
+    cls: "PreTrainedModel", name: str, dduf_entries: Dict[str, DDUFEntry], **kwargs
+) -> "PreTrainedModel":
+    """
+    Load a transformers model from a DDUF archive.
+
+    In practice, `transformers` do not provide a way to load a model from a DDUF archive. This function is a workaround
+    by instantiating a model from the config file and loading the weights from the DDUF archive directly.
+    """
+    config_file = dduf_entries.get(f"{name}/config.json")
+    if config_file is None:
+        raise EnvironmentError(
+            f"Could not find a config.json file for component {name} in DDUF file (contains {dduf_entries.keys()})."
+        )
+    config = PretrainedConfig(**json.loads(config_file.read_text()))
+
+    model = cls(config)
+    weight_files = [
+        entry
+        for entry_name, entry in dduf_entries.items()
+        if entry_name.startswith(f"{name}/") and entry_name.endswith(".safetensors")
+    ]
+
+    if not weight_files:
+        raise EnvironmentError(
+            f"Could not find any weight file for component {name} in DDUF file (contains {dduf_entries.keys()})."
+        )
+    if not is_safetensors_available():
+        raise EnvironmentError(
+            "Safetensors is not available, cannot load model from DDUF. Please `pip install safetensors`."
+        )
+
+    # TODO: use kwargs (torch_dtype, device_map, max_memory, offload_folder, offload_state_dict, use_safetensors, low_cpu_mem_usage)
+    # TODO: is there something else we should take care of?
+    for entry in weight_files:
+        with entry.as_mmap() as mm:
+            state_dict = safetensors.torch.load(mm)
+            model.load_state_dict(state_dict)
```

Comment:
mentioned by @LysandreJik better to use `model.from_pretrained(state_dict)`

Reply:
Indeed, this won't work with quantization otherwise as we perform some changes to the model.

Reply:
100 percent should be replaced with `from_pretrained()`.

Reply:
Suggested a solution here: https://github.com/huggingface/diffusers/pull/10171#issuecomment-2532095038. Waiting for confirmation before moving forward